src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.Linux.cs

```diff
@@ -328,6 +337,46 @@ protected ulong ClearRegister
             set {  *(ulong*)(_registerViewPointer->GPCLR) = value; }
         }
 
+        /// <summary>
+        /// Returns the peripheral base address on the CPU bus of the raspberry pi based on the ranges set within the device tree.
+        /// </summary>
+        /// <remarks>
+        /// The range examined in this method is essentially a mapping between where the peripheral base address on the videocore bus and its
+        /// address on the cpu bus. The return value is 32bit (is in the first 4GB) even on 64 bit operating systems (debian / ubuntu tested) but may change in the future
+        /// This method is based on bcm_host_get_peripheral_address() in libbcm_host which may not exist in all linux distributions.
+        /// </remarks>
+        /// <returns>This returns the peripheral base address as a 32 bit address or 0xFFFFFFFF when in error.</returns>
+        private uint GetPeripheralBaseAddress()
+        {
+            uint cpuBusPeripheralBaseAddress = InvalidPeripheralBaseAddress;
+            uint vcBusPeripheralBaseAddress;
+
+            using (BinaryReader rdr = new BinaryReader(File.Open(DeviceTreeRanges, FileMode.Open, FileAccess.Read)))
+            {
+                // get the Peripheral Base Address on the VC bus from the device tree this is to be used to verify that
+                // the right thing is being read and should always be 0x7E000000
+                vcBusPeripheralBaseAddress = BinaryPrimitives.ReadUInt32BigEndian(rdr.ReadBytes(4));
+
+                // get the Peripheral Base Address on the CPU bus from the device tree.
+                cpuBusPeripheralBaseAddress = BinaryPrimitives.ReadUInt32BigEndian(rdr.ReadBytes(4));
+
+                // if the CPU bus Peripheral Base Address is 0 then assume that this is a 64 bit address and so read the next 32 bits.
```

Comment:
double checking - should you be reading next 32 or 64 bits?

Reply:
@krwq 

The range for the Pi4 has the address defined as a big-endian 64 bit value with the first 32bits set to zero is it will allways be in the first 4GB.

`ranges = <0x7e000000  0x0 0xfe000000  0x01800000>,`

This is a 32bit address for the videocore pba, then a 64bit address for the userland pba and then a 32bit value for the pba length.

Whereas the Pi3 has a 32 bit value  for the userland PBA  (and similarly for the Pi2)

`ranges = <0x7e000000 0x3f000000 0x1000000>,`

So this code and also the libbcm_host userland code tests for 64 bit by reading the first 32bit value of the pba and then if it is zero knows that it is 64 bits wide and so reads the next 32 bit value for the pba.

 Note that early versions of Raspbian buster have the correct device tree but outdated libbcm_host and so a zero is returned. This code detects that and uses the internal method to determine the pba which is correct.